### PR TITLE
Baseline support

### PR DESCRIPTION
**Pull request by woehrl01**
Added baseline support (see #132)

You have the ability for a custom baseline function (```float(*YGBaselineFunc)(YGNodeRef node);```) to return whatever baseline you want.

### CODE REVIEW

**Comment by woehrl01** [comment]
Not sure if we should cache the baseline, as we get the baseline twice. First to get the maximum and second to position it. 

I'm thinking about something like ```computedFlexBasisGeneration``` just named ```baselineCalculationGeneration``` and a cached value property.

File: yoga/Yoga.c [hunk 1]
```diff
@@ -334,6 +335,20 @@ YGMeasureFunc YGNodeGetMeasureFunc(const YGNodeRef node) {
   return node->measure;
 }
 
+
+void YGNodeSetBaselineFunc(const YGNodeRef node, YGBaselineFunc baselineFunc) {
+  if (baselineFunc == NULL) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: `node->baseline = baselineFunc;` is enough

File: yoga/Yoga.c [hunk 2]
```diff
@@ -941,6 +956,61 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
...
+
+static float YGBaselineOfFirstLine(const YGNodeRef node, const YGFlexDirection mainAxis, const float parentWidth)
+{
+  if(node->baseline != NULL)
+  {
+    return node->baseline(node);
+  }
+
+  YGNodeRef baselineChild = NULL;
+  for(unsigned int i = 0; i < YGNodeGetChildCount(node); ++i)
```
Comment (emilsjolander) [inline yoga/Yoga.c]: uint32_t

Comment (emilsjolander) [inline yoga/Yoga.c]: i++

**Comment by woehrl01** [comment]
> Please run format.sh in the mean time.

I'd love to do this, but the effort to create the environment to run this on windows, is currently too much work for me. As it seams I need a full linux runtime to run this.

**Comment by woehrl01** [comment]
@emilsjolander clang to the rescue! 👍 strangely clang-format on windows doesn't allow wildcard. So I hope I didn't miss a file.

**Comment by woehrl01** [comment]
> most likely your shell is not doing the expansion.

jep 😄

File: yoga/Yoga.c [hunk 3]
```diff
@@ -2452,19 +2494,30 @@ static void YGNodelayoutImpl(const YGNodeRef node,
 
       // compute the line's height and find the endIndex
       float lineHeight = 0;
+      float maxAscentForCurrentLine = 0;
```
Comment (emilsjolander) [inline yoga/Yoga.c]: Declare these further down where they are used.

File: yoga/Yoga.c [hunk 4]
```diff
@@ -941,6 +950,39 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
   }
 }
 
+static float YGBaselineOfFirstLine(const YGNodeRef node, const YGFlexDirection crossAxis) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: `YGNodeBaseline`

File: yoga/Yoga.c [hunk 5]
```diff
@@ -2452,19 +2494,30 @@ static void YGNodelayoutImpl(const YGNodeRef node,
...
           if (child->lineIndex != i) {
             break;
           }
-
           if (YGNodeIsLayoutDimDefined(child, crossAxis)) {
             lineHeight = fmaxf(lineHeight,
                                child->layout.measuredDimensions[dim[crossAxis]] +
                                    YGNodeMarginForAxis(child, crossAxis, availableInnerWidth));
           }
+          if (performLayout && YGNodeAlignItem(node, child) == YGAlignBaseline) {
```
Comment (woehrl01) [inline yoga/Yoga.c]: isn't performlayout checked at line 2526 a little bit down below?

File: gentest/fixtures/YGAlignBaseline.html [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    <div style="width: 50px; height: 20px;"></div>
+  </div>
+  <div style="width: 50px; height: 20px;">
+    <div style="width: 50px; height: 15px;"></div>
+  </div>
+</div>
+
+
+<div id="align_baseline_column" style="width: 100px; height: 100px; flex-direction:column; align-items: baseline;">
+  <div style="width: 50px; height: 50px;"></div>
```
Comment (emilsjolander) [inline gentest/fixtures/YGAlignBaseline.html]: please test with different widths on the two items as width is the axis being aligned here

File: gentest/fixtures/YGAlignBaseline.html [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+<div id="align_baseline_multiline_column" style="width: 100px; height: 100px; flex-direction:column; align-items: baseline;flex-wrap:wrap;">
+  <div style="width: 50px; height: 50px;"></div>
+  <div style="width: 50px; height: 50px;">
+    <div style="width: 50px; height: 10px;"></div>
+  </div>
+  <div style="width: 50px; height: 20px;">
+    <div style="width: 50px; height: 10px;"></div>
+  </div>
+    <div style="width: 50px; height: 20px;"></div>
```
Comment (emilsjolander) [inline gentest/fixtures/YGAlignBaseline.html]: indentation

File: gentest/fixtures/YGAlignBaseline.html [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+<div id="align_baseline_multiline_row_and_column" style="width: 100px; height: 100px; flex-direction:row; align-items: baseline;flex-wrap:wrap;">
+  <div style="width: 50px; height: 50px;"></div>
+  <div style="width: 50px; height: 50px;flex-direction:column;">
+    <div style="width: 50px; height: 10px;"></div>
+  </div>
+  <div style="width: 50px; height: 20px;flex-direction:column;">
+    <div style="width: 50px; height: 10px;"></div>
+  </div>
+    <div style="width: 50px; height: 20px;"></div>
```
Comment (emilsjolander) [inline gentest/fixtures/YGAlignBaseline.html]: indendation

File: gentest/fixtures/YGAlignBaseline.html [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+<div id="align_baseline_self" style="width: 100px; height: 100px; flex-direction:row;">
+  <div style="width: 50px; height: 50px;align-self: baseline;"></div>
+  <div style="width: 50px; height: 20px;align-self: baseline;">
+    <div style="width: 50px; height: 10px;"></div>
+  </div>
+</div>
+
+<div id="align_baseline_multiline_column" style="width: 100px; height: 100px; flex-direction:column; align-items: baseline;flex-wrap:wrap;">
+  <div style="width: 50px; height: 50px;"></div>
```
Comment (emilsjolander) [inline gentest/fixtures/YGAlignBaseline.html]: please vary the widths as this is a column layout

File: gentest/fixtures/YGAlignBaseline.html [hunk 10]
```diff
@@ -0,0 +1,87 @@
+<div id="align_baseline" style="width: 100px; height: 100px; flex-direction:row; align-items: baseline;">
```
Comment (emilsjolander) [inline gentest/fixtures/YGAlignBaseline.html]: This whole file would probably make sense to have as part of the YGAlignItems.html + YGAlignSelf.html files which already exist

File: yoga/Yoga.c [hunk 11]
```diff
@@ -929,7 +938,12 @@ static inline float YGNodePaddingAndBorderForAxis(const YGNodeRef node,
 }
 
 static inline YGAlign YGNodeAlignItem(const YGNodeRef node, const YGNodeRef child) {
-  return child->style.alignSelf == YGAlignAuto ? node->style.alignItems : child->style.alignSelf;
+  YGAlign align =
```
Comment (emilsjolander) [inline yoga/Yoga.c]: const

File: yoga/Yoga.c [hunk 12]
```diff
@@ -941,6 +955,41 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
   }
 }
 
+static float YGBaseline(const YGNodeRef node,
+                        const YGFlexDirection crossAxis) {
+  if (node->baseline != NULL) {
+    const float baseline = node->baseline(node);
+    if (YGFloatIsUndefined(baseline)) {
+      return node->layout.measuredDimensions[dim[crossAxis]];
```
Comment (emilsjolander) [inline yoga/Yoga.c]: This will always be height as baseline is undefined for column flex direction. right? in that case I think it is better to be clear and use `YGDimensionHeight` instead of `dim[crossAxis]`

File: yoga/Yoga.c [hunk 13]
```diff
@@ -941,6 +955,41 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
   }
 }
 
+static float YGBaseline(const YGNodeRef node,
+                        const YGFlexDirection crossAxis) {
+  if (node->baseline != NULL) {
+    const float baseline = node->baseline(node);
+    if (YGFloatIsUndefined(baseline)) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: Is there a good reason to allow undefined return value for baseline? I would prefer asserting and crashing as it would probably indicate a bug. What do you think?

File: yoga/Yoga.c [hunk 14]
```diff
@@ -941,6 +955,41 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
...
+      return node->layout.measuredDimensions[dim[crossAxis]];
+    }
+    return baseline;
+  }
+
+  YGNodeRef baselineChild = NULL;
+  for (uint32_t i = 0; i < YGNodeGetChildCount(node); i++) {
+    const YGNodeRef child = YGNodeGetChild(node, i);
+    if (child->style.positionType == YGPositionTypeAbsolute || child->lineIndex > 0) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: Why `child->lineIndex > 0`? I understand that it skips multiline children but i'm not sure why.

File: yoga/Yoga.c [hunk 15]
```diff
@@ -941,6 +955,41 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
...
+      break;
+    }
+
+    if (baselineChild == NULL) {
+      baselineChild = child;
+    }
+  }
+
+  if (baselineChild == NULL) {
+    return node->layout.measuredDimensions[dim[crossAxis]];
```
Comment (emilsjolander) [inline yoga/Yoga.c]: same as above regarding `dim[crossAxis]`

File: yoga/Yoga.c [hunk 16]
```diff
@@ -941,6 +955,41 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
...
+  }
+
+  YGNodeRef baselineChild = NULL;
+  for (uint32_t i = 0; i < YGNodeGetChildCount(node); i++) {
+    const YGNodeRef child = YGNodeGetChild(node, i);
+    if (child->style.positionType == YGPositionTypeAbsolute || child->lineIndex > 0) {
+      continue;
+    }
+
+    if (YGNodeAlignItem(node, child) == YGAlignBaseline) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: The baseline of a container is defined by its first baseline aligned child, but when there is no baseline aligned child then the container's baseline is defined by it's last child. This seems odd. Are you sure we have test cases covering edge cases here to make sure this is how it works on the web?

File: yoga/Yoga.c [hunk 17]
```diff
@@ -941,6 +955,41 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
...
+      baselineChild = child;
+    }
+  }
+
+  if (baselineChild == NULL) {
+    return node->layout.measuredDimensions[dim[crossAxis]];
+  }
+
+  const float baseline = YGBaseline(baselineChild, crossAxis);
+  return baseline + baselineChild->layout.position[pos[crossAxis]];
```
Comment (emilsjolander) [inline yoga/Yoga.c]: same as above regarding `pos[crossAxis]`

File: yoga/Yoga.c [hunk 18]
```diff
@@ -966,6 +1015,21 @@ static inline bool YGNodeIsFlex(const YGNodeRef node) {
...
+  }
+
+  for (uint32_t i = 0; i < YGNodeGetChildCount(node); i++) {
+    const YGNodeRef child = YGNodeGetChild(node, i);
+    if (child->style.positionType == YGPositionTypeRelative &&
+        child->style.alignSelf == YGAlignBaseline) {
+      return true;
+    }
+  }
+  return false;
```
Comment (emilsjolander) [inline yoga/Yoga.c]: nit: empty line above

File: yoga/Yoga.h [hunk 19]
```diff
@@ -49,6 +49,9 @@ typedef YGSize (*YGMeasureFunc)(YGNodeRef node,
                                 YGMeasureMode widthMode,
                                 float height,
                                 YGMeasureMode heightMode);
+
+typedef float (*YGBaselineFunc)(YGNodeRef node);
```
Comment (emilsjolander) [inline yoga/Yoga.h]: nit: remove added empty lines

File: yoga/Yoga.c [hunk 20]
```diff
@@ -2452,19 +2516,30 @@ static void YGNodelayoutImpl(const YGNodeRef node,
...
           if (child->lineIndex != i) {
             break;
           }
-
           if (YGNodeIsLayoutDimDefined(child, crossAxis)) {
             lineHeight = fmaxf(lineHeight,
                                child->layout.measuredDimensions[dim[crossAxis]] +
                                    YGNodeMarginForAxis(child, crossAxis, availableInnerWidth));
           }
+          if (YGNodeAlignItem(node, child) == YGAlignBaseline) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: same as above. Don't reference `crossAxis` as it confuses the fact that baseline is only defined for row layouts

File: yoga/Yoga.c (hunk 13 above)
Comment (woehrl01) [inline yoga/Yoga.c]: I'm not sure my self about this. I thought this could be a kind of "feature", so if you return undefined, you simply use the nodes height. But crashing would be fine too for me. What I'm not sure about is if we explicitly add the padding-top here or if the implementation of the custom function needs to consider this.

File: yoga/Yoga.c (hunk 14 above)
Comment (woehrl01) [inline yoga/Yoga.c]: we use only the first line of the children for base layout alignment. At least this is how chrome handles it.

File: yoga/Yoga.c (hunk 13 above)
Comment (emilsjolander) [inline yoga/Yoga.c]: The custom function should not take padding into account. We don't expect this for the measure function so I would like to preserve that here if possible.

Let's crash for now. If we find a valid reason to have this feature we can implement it later.

File: yoga/Yoga.c (hunk 14 above)
Comment (emilsjolander) [inline yoga/Yoga.c]: yes, but this looks like we are skipping over children with multiple lines instead of just looking at their first line?

File: yoga/Yoga.c (hunk 16 above)
Comment (emilsjolander) [inline yoga/Yoga.c]: So the current code is wrong i think. Right? As it will pick the last child in the case no child is baseline aligned?

File: yoga/Yoga.c (hunk 14 above)
Comment (woehrl01) [inline yoga/Yoga.c]: oh, yep. We should break instead of continue here!

File: yoga/Yoga.c (hunk 16 above)
Comment (woehrl01) [inline yoga/Yoga.c]: no as we only use the child if ```baselineChild``` is still ```NULL```.

**Comment by emilsjolander** [comment]
@woehrl01 If you push all updates in one go (Ok if they are different commits) that would make it more clear to me and others when you are finished addressing comments so we can have another look 👍

File: yoga/Yoga.c (hunk 20 above)
Comment (woehrl01) [inline yoga/Yoga.c]: should I use ```YGFlexDirectionColumn``` here on the ```YGNodeLeadingMargin``` and ```YGNodeMarginForAxis``` calls?

**Comment by woehrl01** [comment]
@emilsjolander sorry, will do next time!

File: yoga/Yoga.c (hunk 20 above)
Comment (emilsjolander) [inline yoga/Yoga.c]: yes :) and replace `dim[crossAxis]`

**Comment by woehrl01** [comment]
I think I got all your points, please have a look 😄

File: yoga/Yoga.h [hunk 21]
```diff
@@ -49,6 +49,8 @@ typedef YGSize (*YGMeasureFunc)(YGNodeRef node,
                                 YGMeasureMode widthMode,
                                 float height,
                                 YGMeasureMode heightMode);
+typedef float (*YGBaselineFunc)(YGNodeRef node);
+
```
Comment (emilsjolander) [inline yoga/Yoga.h]: nit: remove added empty line

File: yoga/Yoga.c [hunk 22]
```diff
@@ -2501,6 +2574,12 @@ static void YGNodelayoutImpl(const YGNodeRef node,
                 //                (auto) crossAxis dimension.
                 break;
               }
+              case YGAlignBaseline: {
+                child->layout.position[pos[crossAxis]] =
```
Comment (emilsjolander) [inline yoga/Yoga.c]: Could `pos[crossAxis]` be hardcoded? Same with `crossAxis` a couple lines below

File: yoga/Yoga.c (hunk 18 above)
Comment (emilsjolander) [inline yoga/Yoga.c]: Please add this 👍

File: yoga/Yoga.c [hunk 23]
```diff
@@ -941,6 +955,40 @@ static inline YGDirection YGNodeResolveDirection(const YGNodeRef node,
...
+  if (node->baseline != NULL) {
+    const float baseline = node->baseline(node);
+    YG_ASSERT(!YGFloatIsUndefined(baseline), "Expect custom baseline function to not return NaN")
+    return baseline;
+  }
+
+  YGNodeRef baselineChild = NULL;
+  for (uint32_t i = 0; i < YGNodeGetChildCount(node); i++) {
+    const YGNodeRef child = YGNodeGetChild(node, i);
+    if (child->lineIndex > 0) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: not quite sure why we break if a child has more than one line. That would mean if the first child of this node has two lines then the baseline would default to the height of the node. I'm not sure why that would be correct.

Comment (woehrl01) [inline yoga/Yoga.c]: we break as only childs on the first line which have align baseline are taken into account. I already added a testcase for this.

File: yoga/Yoga.c (hunk 18 above)
Comment (woehrl01) [inline yoga/Yoga.c]: not sure what you mean, should I add a empty line above ```return false```?

File: yoga/Yoga.c (hunk 22 above)
Comment (woehrl01) [inline yoga/Yoga.c]: sure, I just wanted to be the same like the others, to not cause confusion. But I'll replace them, to be explicit that it is only used on ```YGEdgeTop```

File: yoga/Yoga.c (hunk 23 above)
Comment (woehrl01) [inline yoga/Yoga.c]: oh, I see the confusion, ```child->lineIndex``` does not mean that it has more than one line. It means it is itself on the seconds (or later) line.

Comment (emilsjolander) [inline yoga/Yoga.c]: @woehrl01 yeah that was my confusion, i'm not super familiar with the multi line code and read it too quickly 🐱

**Comment by woehrl01** [comment]
@emilsjolander I additionally modified ```YGIsBaselineLayout``` to shortcut to false if it's a column layout.

**Comment by anql** [comment]
@woehrl01 The baseline feature is really awesome,  but I'm not sure how to implement the function correctly. Would you give me some suggestions?  Thanks.

**Comment by woehrl01** [comment]
@anql: Sure! You must return the vertical position inside the node you want to align. 

If you do not have a baseline function set, it uses the nodes height. That means that all the nodes are align at their bottom line. 

If you have a text node where you want to align your nodes across each others, you need to return the position of the baseline of the firstline inside the node.

I unpacked my awesome paint skills, to make this a bit clearer: 

image

If the black numbers are the height of the nodes, you need to return the red values, to align the nodes across. If you have multiline text you need to return the value of the first line.

**Comment by anql** [comment]
The paint is so clearer, it's very helpful for me.  👍 @woehrl01
